File: d04.go
package main

import (
	"bufio"
	"github.com/dbabiak/aoc-2020/utils"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type Passport struct {
	byr	int
	iyr	int
	eyr	int
	hgt string
	hcl string
	ecl string
	pid string
	cid string
}


func parsePassport(kvs map[string]string) *Passport {
	// do we want this to be dynamic?
	passport := Passport{}
	for k, v := range kvs {
		switch k {
		case "byr":
			n, err := strconv.Atoi(v)
			utils.Check(err)
			passport.byr = n

		case "iyr":
			n, err := strconv.Atoi(v)
			utils.Check(err)
			passport.iyr = n

		case "eyr":
			n, err := strconv.Atoi(v)
			utils.Check(err)
			passport.eyr = n

		case "hgt": passport.hgt = v
		case "hcl": passport.hcl = v
		case "ecl": passport.ecl = v
		case "pid": passport.pid = v
		case "cid": passport.cid = v
		}
	}
	return &passport
}
var r = regexp.MustCompile("([a-zA-Z]+):([a-zA-Z]+)")

func main() {
	f, err := os.Open("./data/d04.txt")
	utils.Check(err)
	scan := bufio.NewScanner(f)
	for i := 0 ; scan.Scan(); i++ {
		line := scan.Text()
		xs := strings.Split(line, "\n")
		for _, x := range xs {
			kv := r.FindStringSubmatch(x)
			k, v := kv[0], kv[1]
		}
	}
}

File: d04.py
import os
import re
from pprint import pprint
from typing import *

os.system('head data/d04.txt')

print('\n-------------------------------\n')

re.compile(r'([a-zA-Z]+):([a-zA-Z]+)')

def main() -> None:
  with open('data/d04.txt') as fp:
      rows = []
      row = None 
      for i, line in enumerate(map(str.strip, fp)):
          row = row or []
          if line:
            for kv in line.split(' '):
              k, v = kv.split(':')
              row.append((k,v))
          else:
            rows.append(dict(row))
            row = []
  print('part1', part1(rows)) 
  print('part2', part2(rows)) 


REQUIRED_FIELDS = [
  'byr',
  'iyr',
  'eyr',
  'hgt',
  'hcl',
  'ecl',
  'pid',
]

def is_valid(row) -> bool:
  return all(k in row for k in REQUIRED_FIELDS)


def part1(rows: List[Dict]) -> int:
    return sum(
        all(k in row for k in REQUIRED_FIELDS)
        for row in rows
    )



def part2(passports: List[Dict]) -> int:
    predicates = {
      'byr': lambda x: 1920 <= int(x) <= 2002,
      'iyr': lambda x: 2010 <= int(x) <= 2020,
      'eyr': lambda x: 2020 <= int(x) <= 2030,
      'hgt': lambda x: (
           (x.endswith('cm') and (150 <= int(x[:-2]) <= 193))
        or (x.endswith('in') and ( 59 <= int(x[:-2]) <=  76))
      ),
      'hcl': lambda x: re.match('^#[0-9a-f]{6}$', x),
      'ecl': lambda x: x in ('amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'),
      'pid': lambda x: re.match(r'^[0-9]{9}$', x),
      'cid': lambda _: True,
    }
    return sum(
      (
            all(k in passport for k in REQUIRED_FIELDS) 
        and all(predicates[k](v) for k, v in passport.items())
      )
      for passport in passports
    )


          
if __name__ == '__main__':
    main()
